text/3735-field-projections.md

```diff
@@ -0,0 +1,1744 @@
+with simpler syntax:
+- `&raw mut (*ptr).foo` is turned into `ptr->foo`
+- using `NonNull<T>` as a replacement for `*mut T` becomes a lot better when accessing fields,
+
+There is of course a cost associated with introducing a new operator along with the concept of field
+projections.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Implementation Details
+
+In order to facilitate field projections, several interlinked concepts have to be introduced. These
+concepts are:
+
+- [field types],
+  - `Field` traits,
+  - `field_of!` macro,
+  - [`#[projecting]`](#projecting-attribute) attribute
+- projection operator `->`,
+  - `Project` trait,
+  - `Projectable` trait,
+
+To ease understanding, here is a short explanation of the interactions between these concepts. The
+following subsections explain them in more detail, so refer to them in cases of ambiguity. The
+projection operator `->` is governed by the `Project` trait that has `Projectable` as a super trait.
+`Projectable` helps to select the struct whose fields are used for projection. Field types store
+information about a field (such as the base struct and the field type) via the `UnalignedField`
+trait and the `field_of!` macro makes it possible to name the [field types]. Finally, the
+[`#[projecting]`](#projecting-attribute) attribute allows `repr(transparent)` structs to be ignored
+when looking for fields for projection.
+
+### Field Types
+[field type]: #field-types
+[field types]: #field-types
+
+The compiler generates a compiler-internal type for every sized[^2] field of every struct. These types
+can only be named via the `field_of!` macro that has the same syntax as `offset_of!`. Only fields
+accessible to the current scope can be projected. These types are called *field types*.
+
+[^2]: This restriction can be lifted in the future to include unsized types with statically known
+    alignment, but that would have to be done in unison with adding support for those fields in
+    `offset_of!`.
+
+Field types implement the `UnalignedField` trait:
+
+```rust
+/// # Safety
+///
+/// In any instance of the type `Self::Base`, at byte offset `Self::OFFSET`, there exists a
+/// (possibly misaligned) field of type `Self::Type`.
+pub unsafe trait UnalignedField {
+    type Base: ?Sized;
+    type Type: ?Sized;
+
+    const OFFSET: usize;
+}
+```
+
+In the implementation of this trait, `Base` is set to the struct that the field is part of and
+`Type` is set to the type of the field. `OFFSET` is set to the offset in bytes of the field in
+the struct (i.e. `OFFSET = offset_of!(Base, ident)`).
+
+For aligned fields (such as all fields of non-`#[repr(packed)]` structs), their field types also
+implement the `Field` trait:
+
+```rust
+/// # Safety
+///
+/// In any well-aligned instance of the type `Self::Base`, at byte offset `Self::OFFSET`, there
+/// exists a well-aligned field of type `Self::Type`.
+pub unsafe trait Field: UnalignedField {}
```

Comment:
Bikeshed: maybe the other way around would be better, i.e. `Field` is any field and then `AlignedField: Field` is a restriction.

Reply:
@tgross35 From an ergonomic PoV most of the time people would want a well-formed reference, so the name of the common case should be short. Accidentally using the base trait bound meaning you may accidentally allow `a->b` on an unaligned field and causing UB (e.g. creating an unaligned reference).